iot_config/src/lora_field.rs

```diff
@@ -0,0 +1,516 @@
+pub type DevAddrField = LoraField<8>;
+pub type EuiField = LoraField<16>;
+
+pub mod proto {
+    pub use helium_proto::services::iot_config::{DevaddrRangeV1, EuiV1, OrgV1};
+}
+
+#[derive(Clone, Debug, Deserialize, Eq, PartialEq, Serialize)]
+pub struct DevAddrRange {
+    #[serde(alias = "lower")]
+    pub start_addr: DevAddrField,
+    #[serde(alias = "upper")]
+    pub end_addr: DevAddrField,
+}
+
+impl DevAddrRange {
+    pub fn new(
+        start_addr: DevAddrField,
+        end_addr: DevAddrField,
+    ) -> Result<Self, DevAddrRangeError> {
+        if end_addr < start_addr {
+            return Err(DevAddrRangeError::EndLessThanStart);
+        }
+
+        Ok(Self {
+            start_addr,
+            end_addr,
+        })
+    }
+
+    pub fn next_start(&self) -> Result<DevAddrField, DevAddrRangeError> {
+        let end: u64 = self.end_addr.into();
+        Ok(devaddr(end + 1))
+    }
+
+    pub fn contains(&self, range: &Self) -> bool {
+        self.start_addr <= range.start_addr && self.end_addr >= range.end_addr
+    }
+}
+
+#[derive(Clone, Debug, Deserialize, Eq, PartialEq, Serialize)]
+pub struct Eui {
+    pub app_eui: EuiField,
+    pub dev_eui: EuiField,
+}
+
+impl Eui {
+    pub fn new(app_eui: EuiField, dev_eui: EuiField) -> Self {
+        Self { app_eui, dev_eui }
+    }
+}
+
+#[derive(thiserror::Error, Debug)]
+pub enum ParseError {
+    #[error("char len mismatch: expected {0}, found {1}")]
+    LengthMismatch(usize, usize),
+    #[error("parse int failed: {0}")]
+    ParseInt(#[from] std::num::ParseIntError),
+}
+
+#[derive(thiserror::Error, Debug)]
+#[error("net_id field error: {0}")]
+pub struct NetIdError(#[from] ParseError);
+
+#[derive(thiserror::Error, Debug)]
+#[error("devaddr field error: {0}")]
+pub struct DevAddrError(#[from] ParseError);
+
+#[derive(thiserror::Error, Debug)]
+#[error("eui field error: {0}")]
+pub struct EuiError(#[from] ParseError);
+
+#[derive(thiserror::Error, Debug)]
+#[error("invalid net id type: {0}")]
+pub struct InvalidNetId(u32);
+
+#[derive(thiserror::Error, Debug)]
+pub enum DevAddrRangeError {
+    #[error("devaddr end less than start")]
+    EndLessThanStart,
+    #[error("devaddr next addr failed")]
+    NextStartUnavailable,
+}
+
+impl<const WIDTH: usize> PartialOrd for LoraField<WIDTH> {
+    fn partial_cmp(&self, other: &Self) -> Option<std::cmp::Ordering> {
+        self.0.partial_cmp(&other.0)
+    }
+}
+
+impl<const WIDTH: usize> From<LoraField<WIDTH>> for u64 {
+    fn from(field: LoraField<WIDTH>) -> Self {
+        field.0
+    }
+}
+
+impl<const WIDTH: usize> From<LoraField<WIDTH>> for i64 {
+    fn from(field: LoraField<WIDTH>) -> Self {
+        field.0 as i64
+    }
+}
+
+impl<const WIDTH: usize> From<LoraField<WIDTH>> for u32 {
+    fn from(field: LoraField<WIDTH>) -> Self {
+        field.0 as u32
+    }
+}
+
+impl<const WIDTH: usize> From<LoraField<WIDTH>> for i32 {
+    fn from(field: LoraField<WIDTH>) -> Self {
+        field.0 as i32
+    }
+}
```

Comment:
Would it make sense to only provide `64` conversions for `EuiField`, `32` conversions for `DevAddrField` and `NetIdField`?

In hopes if disallowing conversions that would remove necessary bits.

Reply:
That’s a good call; I think most of those were a holdover from morphing db requirements that are no longer relevant

Reply:
fixed [4c94984](https://github.com/helium/oracles/pull/287/commits/4c949841518190a56205e8e3ccb88b1138ed09b4)